_posts/2021-02-16-message-passing.md

```diff
@@ -0,0 +1,209 @@
+
+`receive` searches for matching messages in the message queue (in the order
+they arrived), or waits for new messages if none were found. Searching through
+the message rather than signal queue means it doesn't have to contend with
+processes that send messages, which greatly increases performance.
+
+This ability to "receive" specific messages is very convenient: we're not
+always in a context where we can decide what to do with a message and having to
+manually lug around all unhandled messages is certainly annoying.
+
+Unfortunately, sweeping the search under the rug doesn't make it go away:
+
+```erlang
+receive
+    {reply, Result} ->
+        {ok, Result}
+end
+```
+
+The above expression finishes instantly if the next message in the queue
+matches `{reply, Result}`, but if there's no matching message it has to walk
+through them all before giving up. This is expensive when there are a lot of
+messages queued up which is common for server-like processes, and since
+`receive` expressions can match on just about anything there's little that can
+be done to optimize the search itself.
+
+The only optimization we do at the moment is to mark a starting point for the
+search when we know that a message couldn't exist prior to a certain point.
+Let's revisit the request-response idiom:
+
+```erlang
+Mref = monitor(process, Pid),
+Pid ! {self(), Mref, Request},
+receive
+    {Mref, Response} ->
+        erlang:demonitor(Mref, [flush]),
+        {ok, Response};
+    {'DOWN', Mref, _, _, Reason} ->
+        {error, Reason}
+end
+```
+
+Since the reference created by `monitor/2` is globally unique and cannot exist
+before said call, and the `receive` only matches messages that contain said
+reference, we don't need to look at any of the messages received before then.
+
+This makes the idiom efficient even on processes that have absurdly long
+message queues, but unfortunately it isn't something we can do in the general
+case. While you as a programmer can be sure that a certain response must come
+after its request even without a reference, for example by using your own
+sequence numbers, the compiler can't read your intent and has to assume that
+you want _any_ message that matches.
+
+Figuring out whether the above optimization has kicked in is rather annoying at
+the moment. It requires inspecting BEAM assembly and even then you're not
+guaranteed that it will work due to some annoying limitations:
+
+1. We only support one message position at a time: a function that creates a
+   reference, calls another function that uses this optimization, and then
+   returns to `receive` with the first reference, will end up searching through
+   the entire message queue.
+1. It only works within a single function clause: both reference creation and
+   `receive` need to be next to each other and you can't have multiple
+   functions calling a common `receive` helper.
+
+We've addressed these shortcomings in the upcoming OTP 24 release, and have
```

Comment:
Maybe mention that the creation of the ref and the receive using the ref must be in the same module, otherwise the compiler can't find and optimize it. So there is still a limitation, but much less than before

Reply:
They don't have to be the same module: it can cross module boundaries in some cases and we use this to ensure that `demonitor(MRef, [flush])` is optimized properly. :sunglasses: 

The specifics are rather intricate and would go a bit beyond the scope of the post, so I left it unsaid knowing that the `recv_opt_info` flag will tell you when it happens.